File: src/main/java/org/example/bankramenserver/domain/recurring/presentation/dto/response/RecurringPaymentResponse.java
package org.example.bankramenserver.domain.recurring.presentation.dto.response;

import org.example.bankramenserver.domain.recurring.domain.RecurringPayment;

import java.time.LocalDate;
import java.util.UUID;

public record RecurringPaymentResponse(
        UUID recurringPaymentId,
        String name,
        Long amount,
        String category,
        String categoryDisplayName,
        RecurringPayment.Cycle cycle,
        int billingDay,
        LocalDate nextBillingDate,
        RecurringPayment.RegistrationType registrationType,
        boolean confirmed,
        int transactionCount,
        LocalDate lastPaidDate
) {
    public static RecurringPaymentResponse from(RecurringPayment recurringPayment) {
        LocalDate lastPaidDate = recurringPayment.getTransactions().stream()
                .map(item -> item.getTransaction().getTransactionDate())
                .max(LocalDate::compareTo)
                .orElse(null);

        var category = recurringPayment.getCategory();

        return new RecurringPaymentResponse(
                recurringPayment.getId(),
                recurringPayment.getName(),
                recurringPayment.getAmount(),
                category != null ? category.name() : null,
                category != null ? category.getDisplayName() : null,
                recurringPayment.getCycle(),
                recurringPayment.getBillingDay(),
                recurringPayment.getNextBillingDate().toLocalDate(),
                recurringPayment.getRegistrationType(),
                recurringPayment.isConfirmed(),
                recurringPayment.getTransactions().size(),
                lastPaidDate
        );
    }
}

File: src/main/kotlin/org/example/bankramenserver/domain/recurring/presentation/dto/response/RecurringPaymentResponse.kt
package org.example.bankramenserver.domain.recurring.presentation.dto.response

import org.example.bankramenserver.domain.recurring.domain.RecurringPayment
import java.time.LocalDate
import java.util.UUID

data class RecurringPaymentResponse(
    val recurringPaymentId: UUID, val name: String, val amount: Long, val category: String?, val categoryDisplayName: String?,
    val cycle: RecurringPayment.Cycle, val billingDay: Int, val nextBillingDate: LocalDate, val registrationType: RecurringPayment.RegistrationType,
    val confirmed: Boolean, val transactionCount: Int, val lastPaidDate: LocalDate?,
) {
    companion object {
        @JvmStatic
        fun from(recurringPayment: RecurringPayment): RecurringPaymentResponse {
            val category = recurringPayment.category
            return RecurringPaymentResponse(
                recurringPayment.id!!, recurringPayment.name, recurringPayment.amount!!, category?.name, category?.displayName,
                recurringPayment.cycle, recurringPayment.billingDay, recurringPayment.nextBillingDate.toLocalDate(), recurringPayment.registrationType,
                recurringPayment.confirmed, recurringPayment.transactions.size, recurringPayment.transactions.maxOfOrNull { it.transaction.transactionDate },
            )
        }
    }
}
